src/policy/v3_policy.cpp

```diff
@@ -0,0 +1,220 @@
+                                          const std::set<Txid>& direct_conflicts,
+                                          int64_t vsize)
+{
+    // Check v3 and non-v3 inheritance.
+    for (const auto& entry : mempool_ancestors) {
+        if (ptx->nVersion != 3 && entry->GetTx().nVersion == 3) {
+            return strprintf("non-v3 tx %s (wtxid=%s) cannot spend from v3 tx %s (wtxid=%s)",
+                             ptx->GetHash().ToString(), ptx->GetWitnessHash().ToString(),
+                             entry->GetSharedTx()->GetHash().ToString(), entry->GetSharedTx()->GetWitnessHash().ToString());
+        } else if (ptx->nVersion == 3 && entry->GetTx().nVersion != 3) {
+            return strprintf("v3 tx %s (wtxid=%s) cannot spend from non-v3 tx %s (wtxid=%s)",
+                             ptx->GetHash().ToString(), ptx->GetWitnessHash().ToString(),
+                             entry->GetSharedTx()->GetHash().ToString(), entry->GetSharedTx()->GetWitnessHash().ToString());
+        }
+    }
+
+    // This function is specialized for these limits, and must be reimplemented if they ever change.
+    static_assert(V3_ANCESTOR_LIMIT == 2);
+    static_assert(V3_DESCENDANT_LIMIT == 2);
+
+    // The rest of the rules only apply to transactions with nVersion=3.
+    if (ptx->nVersion != 3) return std::nullopt;
+
+    // Check that V3_ANCESTOR_LIMIT would not be violated, including both in-package and in-mempool.
+    if (mempool_ancestors.size() + 1 > V3_ANCESTOR_LIMIT) {
+        return strprintf("tx %s (wtxid=%s) would have too many ancestors",
+                         ptx->GetHash().ToString(), ptx->GetWitnessHash().ToString());
+    }
+
+    // Remaining checks only pertain to transactions with unconfirmed ancestors.
+    if (mempool_ancestors.size() > 0) {
+        // If this transaction spends V3 parents, it cannot be too large.
+        if (vsize > V3_CHILD_MAX_VSIZE) {
+            return strprintf("v3 child tx %s (wtxid=%s) is too big: %u > %u virtual bytes",
+                             ptx->GetHash().ToString(), ptx->GetWitnessHash().ToString(), vsize, V3_CHILD_MAX_VSIZE);
+        }
+
+        // Check the descendant counts of in-mempool ancestors.
+        if (!mempool_ancestors.empty()) {
+            const auto& parent_entry = *mempool_ancestors.begin();
+            // If there are any ancestors, this is the only child allowed. The parent cannot have any
+            // other descendants.
+            const auto& children = parent_entry->GetMemPoolChildrenConst();
+            // Don't double-count a transaction that is going to be replaced. This logic assumes that
+            // any descendant of the V3 transaction is a direct child, which makes sense because a V3
+            // transaction can only have 1 descendant.
+            const bool child_will_be_replaced = !children.empty() &&
+                std::any_of(children.cbegin(), children.cend(),
```

Comment:
I think this is more correct in the case of a reorg? We should accept if it double-spends all siblings.

```suggestion
                std::all_of(children.cbegin(), children.cend(),
```

A case that exercises this stuff in `test_v3_reorg` would be nice too. Try submitting another child when there's more than one child already due to reorg.

Reply:
I don't think so, otherwise this is needlessly making things worse than it needs to be.

Imagine transactions C1 and C2 are both v3, both in the mempool, with confirmed v3 parent P.  C2 can be arbitrarily big, since it entered the mempool with no unconfirmed parents.

P is reorged out, and now P, C1, and C2 are all in the mempool.

Suppose D is an RBF of C1 that would pass our RBF rules.  However, C2 is a big transaction, and being forced to conflict with C2 (eg due to sibling eviction or whatever) might make D fail, even though the mempool would be strictly improved by just letting D replace C1, and leaving C2 in place.

Anyway I don't think this edge case behavior is super important, but if we have to pick something, I think what is implemented here is probably right.

Reply:
I agree that in the case above D shouldn't need to conflict with all children, so will leave as is.

Reply:
I still would like a test for this, maybe as a follow-up

Reply:
@instagibbs To test my understanding of the discussion here I created a test for this case https://github.com/ismaelsadeeq/bitcoin/commit/87fc7f0a8d8ad1fc2af60ee6cc678df5e6fb01dd

I agree we should not force siblings tx to conflict with reorged siblings, since we dont enforce v3 rules after reorg.

Maybe this should be documented in the comment.

Reply:
@ismaelsadeeq that test looks good to me 👍 can I cherry-pick this into #29424?